enhancements/kube-apiserver/stability-point-to-point-network-check.md

```diff
@@ -0,0 +1,418 @@
+be a good place to talk about core concepts and how they relate.
+
+#### RelatedEndpoint Custom Resource
+
+```go
+package v1alpha1
+
+// group: network.openshift.io
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PodNetworkConnectivityCheck
+type PodNetworkConnectivityCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	Spec   PodNetworkConnectivityCheckSpec   `json:"spec"`
+	Status PodNetworkConnectivityCheckStatus `json:"status"`
+}
+
+type PodNetworkConnectivityCheckSpec struct {
+	// SourcePod names the pod from which the condition will be checked
+	SourcePod string `json:"sourcePod"`
+	// EndpointAddress to check. A TCP address of the form host:port. Note that
+	// if host is a DNS name, then the check would fail if the DNS name cannot
+	// be resolved. Specify an IP address for host to bypass DNS name lookup.
+	TargetEndpoint string `json:"targetEndpoint"`
+}
+
+type PodNetworkConnectivityCheckStatus struct {
+	// Log of check events
+	Log []LogEntry `json:"log"`
+	// Conditions summarize the status of the check
+	Conditions []metav1.ConditionStatus `json:"conditions"`
+}
+
+// LogEntry records events
+type LogEntry struct {
+	// Time of occurrence of check action.
+	Time metav1.Time `json:"time,omitempty"`
+	// Status indicates if the action performed by the check was successful or not.
+	Success bool `json:"status"`
+	// Reason for status in a machine readable format.
+	Reason LogEntryReason `json:"reason"`
+	// Message explaining status in a human readable format.
+	Message string `json:"message"`
+	// Latency records how long the action mentioned in the entry took.
+	Latency time.Duration `json:"latency"`
+}
+
+type PodNetworkConnectivityCheckCondition struct {
+	// Type of the condition
+	Type PodNetworkConnectivityCheckConditionType `json:"type"`
+	// Status of the condition
+	Status metav1.ConditionStatus `json:"status"`
+	// machine readable reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+	// human readable message indicating details about last transition.
+	Message string `json:"message,omitempty"`
+	// Last time the condition transit from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+type LogEntryReason string
+
+type PodNetworkConnectivityCheckConditionType string
+
+const (
+	Reachable    PodNetworkConnectivityCheckConditionType = "Reachable"
+	DNSDone      LogEntryReason                           = "DNSDone"
+	DNSError     LogEntryReason                           = "DNSError"
+	ConnectDone  LogEntryReason                           = "ConnectDone"
+	ConnectError LogEntryReason                           = "ConnectError"
+)
+
+```
+
+
+#### Point-to-point Network Check Tool
+
+Runs in a container in the interested pod. Executes check and updates `PodNetworkConnectivityCheck` resources for the pod.
```

Comment:
I was expecting a LOT more discussion here.  I.e. why is this not a binary that is in the base image of all of our components and an exec session triggers it?

Reply:
> I was expecting a LOT more discussion here. I.e. why is this not a binary that is in the base image of all of our components and an exec session triggers it?

It doesn't exist at all today.  Starting with "we can solve this narrow problem here, like this" has a lot more immediate value than, " I will build a grand unified theory of X and everyone will like it".